cli/pkg/deviceauth/authenticator.go

```diff
@@ -0,0 +1,246 @@
+	ExpiresIn               int    `json:"expires_in"`
+	PollingInterval         int    `json:"interval"`
+}
+
+// DeviceVerification represents the response from verifying a device.
+type DeviceVerification struct {
+	DeviceCode              string
+	UserCode                string
+	VerificationURL         string
+	VerificationCompleteURL string
+	CheckInterval           time.Duration
+	ExpiresAt               time.Time
+}
+
+// DeviceAuthenticator performs the authentication flow for logging in.
+type DeviceAuthenticator struct {
+	client   *http.Client
+	BaseURL  *url.URL
+	Clock    clock.Clock
+	ClientID string
+}
+
+// New returns an instance of the DeviceAuthenticator
+func New(authURL string) (*DeviceAuthenticator, error) {
+	baseURL, err := url.Parse(authURL)
+	if err != nil {
+		return nil, err
+	}
+
+	authenticator := &DeviceAuthenticator{
+		client:   http.DefaultClient,
+		BaseURL:  baseURL,
+		Clock:    clock.New(),
+		ClientID: OAuthClientID,
+	}
+
+	return authenticator, nil
+}
+
+// VerifyDevice performs the device verification API calls.
+func (d *DeviceAuthenticator) VerifyDevice(ctx context.Context) (*DeviceVerification, error) {
+	req, err := d.newFormRequest(ctx, "oauth/device_authorization", url.Values{
+		"client_id": []string{d.ClientID},
+		"scope":     []string{"full_account"},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := d.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if _, err = checkErrorResponse(res); err != nil {
+		return nil, err
+	}
+
+	deviceCodeRes := &DeviceCodeResponse{}
+	err = json.NewDecoder(res.Body).Decode(deviceCodeRes)
+	if err != nil {
+		return nil, errors.Wrap(err, "error decoding device code response")
+	}
+
+	checkInterval := time.Duration(deviceCodeRes.PollingInterval) * time.Second
+	if checkInterval == 0 {
+		checkInterval = time.Duration(5) * time.Second
+	}
+
+	expiresAt := d.Clock.Now().Add(time.Duration(deviceCodeRes.ExpiresIn) * time.Second)
+
+	return &DeviceVerification{
+		DeviceCode:              deviceCodeRes.DeviceCode,
+		UserCode:                deviceCodeRes.UserCode,
+		VerificationCompleteURL: deviceCodeRes.VerificationCompleteURI,
+		VerificationURL:         deviceCodeRes.VerificationURI,
+		ExpiresAt:               expiresAt,
+		CheckInterval:           checkInterval,
+	}, nil
+}
+
+// GetAccessTokenForDevice uses the device verification response to fetch an
+// access token.
+func (d *DeviceAuthenticator) GetAccessTokenForDevice(ctx context.Context, v DeviceVerification) (*OAuthTokenResponse, error) {
+	for {
+		// This loop begins right after we open the user's browser to send an
+		// authentication code. We don't request a token immediately because the
+		// has to complete that authentication flow before we can provide a
+		// token anyway.
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(v.CheckInterval):
+			// Ready to check again.
+		}
```

Comment:
Have you checked that doing `ctrl+C` when this is running will still exit? If not, you can solve it with `runtime/pkg/graceful.WithCancelOnTerminate`

Reply:
not checked but thought `ctx.Done()` will be called. Will check

Reply:
Just noticed the context originates from the Cobra command, so I think you're right – probably they handle it